src/imperator/data/events.rs

```diff
@@ -0,0 +1,252 @@
+        let mut expects_scope = Scopes::Country;
+        let mut expects_from_token = key.clone();
+
+        if let Some(event_type) = block.get_field_value("type") {
+            match event_type.as_str() {
+                "character_event" => {
+                    expects_scope = Scopes::Character;
+                    expects_from_token = event_type.clone();
+                }
+                "minor_character_event" => {
+                    expects_scope = Scopes::Character;
+                    expects_from_token = event_type.clone();
+                }
+                "country_event" => expects_from_token = event_type.clone(),
+                "minor_country_event" => expects_from_token = event_type.clone(),
+                "major_country_event" => expects_from_token = event_type.clone(),
+                "state_event" => {
+                    expects_scope = Scopes::State;
+                    expects_from_token = event_type.clone();
+                }
+                "province_event" => {
+                    expects_scope = Scopes::Province;
+                    expects_from_token = event_type.clone();
+                }
+                _ => (),
+            }
+        }
+
+        Self { key, block, expects_scope, expects_from_token }
+    }
+    pub fn validate(&self, data: &Everything) {
+        let mut vd = Validator::new(&self.block, data);
+
+        let mut tooltipped_immediate = Tooltipped::Past;
+        let mut tooltipped = Tooltipped::Yes;
+        if let Some((namespace, _)) = self.key.as_str().split_once('.') {
+            if !data.item_exists(Item::EventNamespace, namespace) {
+                let msg = format!("event file should start with `namespace = {namespace}`");
+                let info = "otherwise the event won't be found in-game";
+                error_info(&self.key, ErrorKey::EventNamespace, &msg, info);
+            }
+        }
+
+        let mut sc = ScopeContext::new(self.expects_scope, &self.expects_from_token);
+        sc.set_strict_scopes(false);
+
+        vd.field_choice("type", EVENT_TYPES);
+        vd.field_bool("hidden");
+        vd.field_bool("fire_only_once");
+        vd.field_target("goto_location", &mut sc, Scopes::Province);
+
+        vd.field_validated_sc("title", &mut sc, validate_desc);
+        vd.field_validated_sc("desc", &mut sc, validate_desc);
+
+        let hidden = self.block.field_value_is("hidden", "yes");
+        if hidden {
+            tooltipped_immediate = Tooltipped::No;
+            tooltipped = Tooltipped::No;
+        }
+
+        if !hidden {
+            vd.req_field("picture");
+        }
+        vd.field_item("picture", Item::EventPicture);
+
+        // TODO - there can be up to 3 left_portrait and 3 right_portrait fields in an event, not sure how to do this
+        vd.field_target("left_portrait", &mut sc, Scopes::Character);
+        vd.field_target("right_portrait", &mut sc, Scopes::Character);
```

Comment:
You can do `self.block.count_keys("left_portrait")` to count them directly and then emit a warning. But what I tend to do instead is use `let mut count = 0`, then `vd.multi_field_validated_key_block`, and in its closure start with `count += 1` and then a check for `count == 4`. That way I can emit a warning exactly on the key of the 4th item, to say that there should be no more than 3.

Since here the logic for `left_portrait` and `right_portrait` is the same and the validator gets kinda long, it would make sense to start with `for field in &["left_portrait", "right_portrait"] {` and then repeat the validator only once.